lustre/client.lua

```diff
@@ -1,30 +1,106 @@
+local cosock = require 'cosock'
 local Request = require "luncheon.request"
 local Response = require "luncheon.response"
 local Handshake = require 'lustre.handshake'
 local Key = require 'lustre.handshake.key'
 local Config = require 'lustre.config'
+local Frame = require 'lustre.frame'
+local FrameHeader = require "lustre.frame.frame_header"
+local OpCode = require "lustre.frame.opcode"
+local CloseCode = require 'lustre.frame.close'
+local Message = require "lustre.message"
 
+--TODO We should obfuscate WebSocketClient from the user. They should
+-- only ever interact with a WebSocket class that acts as a client when 
+-- they call `connect`, and a server when they call `accept`.
 ---@class WebSocketClient
 ---
 ---@field url string the endpoint to hit
 ---@field socket table lua socket
+---@field handshake_key string key used in the websocket handshake
+---@field config Config
+---@field _tx table
+---@field _rx table
 local WebSocketClient = {}
 WebSocketClient.__index = WebSocketClient
 
-function WebSocketClient.new(socket, url, config)
+---Create new client object
+---@param socket table tcp socket to connect on
+---@param url string url to connect
+---@param config Config 
+---@param message_cb function
+---@param error_cb function
+---@param close_cb function
+---@return client WebSocketClient
+---@return err string|nil
+function WebSocketClient.new(socket, url, config, ...)
+    --TODO how was this channel intended to be used again?
+    local _tx, _rx = cosock.channel.new()
     return setmetatable({
         socket = socket,
         url = url or '/',
         handshake_key = Key.generate_key(),
         config = config or Config.default(),
+        _tx = _tx,
+        _rx = _rx,
+        message_cb = message_cb,
+        error_cb = error_cb, 
+        close_cb = close_cb
```

Comment:
These would always be `nil`, if you wanted to get them from the `...` parameter you could do that by wrapping it in a table and then indexing those `extra_args = {...}` and then `message_cb = extra_args[1]`, though it might be nice to implement the builder pattern here, so the `register_*` functions would return `self` this would allow 

```lua
local ws = WebSocketClient.new(sock, '', cfg):register_message_handler(print):register_error_handler(print)
```